packages/crypto/src/keys/ecdsa/utils.ts

```diff
@@ -0,0 +1,211 @@
+import { decodeDer, encodeBitString, encodeInteger, encodeOctetString, encodeSequence } from '../rsa/der.js'
+import { ECDSAPrivateKey as ECDSAPrivateKeyClass, ECDSAPublicKey as ECDSAPublicKeyClass } from './ecdsa.js'
+import { generateECDSAKey } from './index.js'
+import type { Curve } from '../ecdh/index.js'
+import type { ECDSAPublicKey, ECDSAPrivateKey } from '@libp2p/interface'
+
+// 1.2.840.10045.3.1.7 prime256v1 (ANSI X9.62 named elliptic curve)
+const OID_256 = Uint8Array.from([0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07])
+// 1.3.132.0.34 secp384r1 (SECG (Certicom) named elliptic curve)
+const OID_384 = Uint8Array.from([0x06, 0x05, 0x2B, 0x81, 0x04, 0x00, 0x22])
+// 1.3.132.0.35 secp521r1 (SECG (Certicom) named elliptic curve)
+const OID_521 = Uint8Array.from([0x06, 0x05, 0x2B, 0x81, 0x04, 0x00, 0x23])
+
+const P_256_KEY_JWK = {
+  ext: true,
+  kty: 'EC',
+  crv: 'P-256'
+}
+
+const P_384_KEY_JWK = {
+  ext: true,
+  kty: 'EC',
+  crv: 'P-384'
+}
+
+const P_521_KEY_JWK = {
+  ext: true,
+  kty: 'EC',
+  crv: 'P-521'
+}
+
+export function unmarshalECDSAPrivateKey (bytes: Uint8Array): ECDSAPrivateKey {
+  const message = decodeDer(bytes)
+  const privateKey = message[1]
+  const d = uint8ArrayToString(privateKey, 'base64url')
+  const coordinates: Uint8Array = message[2][1][0]
+  let x: string
+  let y: string
+
+  if (privateKey.byteLength === 32) {
+    x = uint8ArrayToString(coordinates.subarray(1, 33), 'base64url')
+    y = uint8ArrayToString(coordinates.subarray(33), 'base64url')
+
+    return new ECDSAPrivateKeyClass({
+      ...P_256_KEY_JWK,
+      key_ops: ['sign'],
+      d,
+      x,
+      y
+    }, {
+      ...P_256_KEY_JWK,
+      key_ops: ['verify'],
+      x,
+      y
+    })
+  }
+
+  if (privateKey.byteLength === 48) {
+    x = uint8ArrayToString(coordinates.subarray(1, 49), 'base64url')
+    y = uint8ArrayToString(coordinates.subarray(49), 'base64url')
+
+    return new ECDSAPrivateKeyClass({
+      ...P_384_KEY_JWK,
+      key_ops: ['sign'],
+      d,
+      x,
+      y
+    }, {
+      ...P_384_KEY_JWK,
+      key_ops: ['verify'],
+      x,
+      y
+    })
+  }
+
+  if (privateKey.byteLength === 66) {
+    x = uint8ArrayToString(coordinates.subarray(1, 67), 'base64url')
+    y = uint8ArrayToString(coordinates.subarray(67), 'base64url')
+
+    return new ECDSAPrivateKeyClass({
+      ...P_521_KEY_JWK,
+      key_ops: ['sign'],
+      d,
+      x,
+      y
+    }, {
+      ...P_521_KEY_JWK,
+      key_ops: ['verify'],
+      x,
+      y
+    })
+  }
+
+  throw new InvalidParametersError(`Private key length was wrong length, got ${privateKey.byteLength}, expected 32`)
+}
+
+export function unmarshalECDSAPublicKey (bytes: Uint8Array): ECDSAPublicKey {
+  const message = decodeDer(bytes)
+  const coordinates = message[1][1][0]
+  let x: string
+  let y: string
+
+  if (coordinates.byteLength === 65) {
+    x = uint8ArrayToString(coordinates.subarray(1, 33), 'base64url')
```

Comment:
Maybe we should move these magic numbers to named constants, it's not immediately obvious why 1,33 should be used.

Reply:
Fixed